### PR TITLE
Update for the procedures for insertion/hot swap of Switch Fabric Module(SFM) by using "config chassis modules shutdown/startup" commands

### PR DESCRIPTION
**Pull request by JunhongMao**
#### Why I did it
For the Nokia SONiC chassis procedures for insertion/hot swap of Switch Fabric Module(SFM), 
the previous solution was using the below commands.
```
sudo nokia_cmd set shutdown-sfm <SFM-Num/Physical-Slot>
```
This PR along with the below PR intend to add the below commands for the equivalent operations.
https://github.com/nokia/sonic-platform/pull/6
```
sudo config chassis modules shutdown/startup <module name>
```

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
1. Add chassis_module_config.py and its service. The service starts up automatically. The example is below.
```
sudo systemctl status chassis-module.service
● chassis-module.service - Chassis module up & down operation
     Loaded: loaded (/lib/systemd/system/chassis-module.service; enabled-runtime; vendor preset: enabled)
     Active: active (running) since Fri 2024-04-05 19:57:25 UTC; 1h 5min ago
   Main PID: 8856 (python3)
      Tasks: 1 (limit: 38314)
     Memory: 16.2M
     CGroup: /system.slice/chassis-module.service
             └─8856 /usr/bin/python3 /usr/local/bin/chassis_module_config.py

Apr 05 19:57:25 ixre-cpm-chassis15 systemd[1]: Started Chassis module up & down operation.
```
3. When the cli command "sudo config chassis modules startup/shutdown" runs, calls chassis_module_set_admin_state.py to do the related operations.
#### How to verify it

```
The below test was carried out on FABRIC-CARD3 module on the supervisor card.
1. Shutdown
sudo config chassis modules shutdown FABRIC-CARD3

2. Check the status to see if the FABRIC-CARD3 was down.
$ show chassis modules status
        Name             Description    Physical-Slot    Oper-Status    Admin-Status       Serial
------------  ----------------------  ---------------  -------------  --------------  -----------
...
FABRIC-CARD3             Unavailable                4          Empty            down          N/A

 
3. Start up the module
sudo config chassis modules startup FABRIC-CARD3

4. Check the status
$ show chassis modules status
        Name             Description    Physical-Slot    Oper-Status    Admin-Status       Serial
------------  ----------------------  ---------------  -------------  --------------  -----------
...
FABRIC-CARD3                    SFM4                4         Online              up  01214400362

5. To test if the operation is still valid when the system reboot. For example, first shut down, 
then after saving config and reboot, the module should keep shutdown status. 
$ sudo config save
Existing files will be overwritten, continue? [y/N]: y

Then check the status to see if the FABRIC-CARD3 was down.
$ show chassis modules status
        Name             Description    Physical-Slot    Oper-Status    Admin-Status       Serial
------------  ----------------------  ---------------  -------------  --------------  -----------
...
FABRIC-CARD3             Unavailable                4          Empty            down          N/A


```

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [x] 202205
- [x] 202211
- [x] 202305

#### Tested branch (Please provide the tested image version)

- [x] 202205

### CODE REVIEW

File: files/build_templates/chassis-module.service [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[Unit]
+Description=Chassis module up & down operation
+ConditionPathExists=/etc/sonic/chassisdb.conf
+Requires=database.service updategraph.service
```
Comment (wumiaont) [inline files/build_templates/chassis-module.service:4]: updategraph.service is a 202205 service which does not exist in 202405 (master). This is for 202205 cherry-pick purpose. We need another PR to change this in master after this PR fix has been picked up by 202205. Any better strategy here?

Comment (JunhongMao) [reply files/build_templates/chassis-module.service:4]: @mlok-nokia , can you comment on this?

**Comment by judyjoseph** [comment]
@JunhongMao I understand that with this PR and https://github.com/nokia/sonic-platform/pull/6, trying to have the shut/start of SFM + swss/syncd processes in the nokia platform submodule.

Can we make this a bit more generic, like when user issue "sudo config chassis modules shutdown FABRIC-CARD3", we can have the implementation in sonic-utilities to start/stop swss/syncd systemd service + call nokia platform API to power up/down the corresponding card ? 

In this way this command will have a sonic common implementation with a platform hook to really power up/down SFM.

**Comment by mlok-nokia** [comment]
> @JunhongMao I understand that with this PR and [nokia/sonic-platform#6](https://github.com/nokia/sonic-platform/pull/6), trying to have the shut/start of SFM + swss/syncd processes in the nokia platform submodule.
> 
> Can we make this a bit more generic, like when user issue "sudo config chassis modules shutdown FABRIC-CARD3", we can have the implementation in sonic-utilities to start/stop swss/syncd systemd service + call nokia platform API to power up/down the corresponding card ?
> 
> In this way this command will have a sonic common implementation with a platform hook to really power up/down SFM.

Hi Judy, The following reasons is why we need to define a service file to subscribe the "CHASSIS_MODULE" tables to shutdown/startup a SFM and related swss/syncd services is that - when users shutdown a SFM and save the config file, then reboot the chassis.  When chassis is booting and loading config, we need to keep the SFM and swss/syncd in the down state based on the configuration. 
Second, number of swss/syncd is associated with a particular SFM module could be different in different Vendor.  That is why we were thinking let the Vendor API to shutdown/startup related swss/syncd  and SFM card is more flexible and straight forward.

Should we have acall to discuss about this?  Thanks.

File: files/scripts/chassis_module_config.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env python3
```
Comment (arlakshm) [inline files/scripts/chassis_module_config.py:1]: why do we another service to handle CLI command?
We can call the platform script from the CLI handler?

Comment (JunhongMao) [reply files/scripts/chassis_module_config.py:1]: In this solution, chassis_module_config.py needs to subscribe the "CHASSIS_MODULE" tables to shutdown/startup a SFM and related swss/syncd services other than directly calls the platform script from the CLI handler. This solution has some considerations, please refer to https://github.com/sonic-net/sonic-buildimage/pull/18578#issuecomment-2045550189. Thanks.

Comment (mlok-nokia) [reply files/scripts/chassis_module_config.py:1]: > why do we another service to handle CLI command? We can call the platform script from the CLI handler?

@arlakshm  For the case --  If user shutdowns the SFM and save the configuration file.  And reboot the whole chassis, do we want to keep the SFM down after the system is up?  If we don't need to handle this case, we can directly call the platform script from the CLI handler.

**Comment by JunhongMao** [comment]
There is a problem: the command "config apply-patch" caused some containers to exit. 

```
admin@ixre-egl-board40:~$ docker ps
CONTAINER ID   IMAGE                             COMMAND                  CREATED       STATUS         PORTS     NAMES
baeabdcd38a9   70dd58e31088                      "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             macsec0
bfe465afc0aa   docker-router-advertiser:latest   "/usr/bin/docker-ini…"   2 hours ago   Up 2 minutes             radv
b710a81b05b9   70dd58e31088                      "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             macsec1
3a8ddbc204ad   docker-fpm-frr:latest             "/usr/bin/docker_ini…"   2 hours ago   Up 2 minutes             bgp0
e553479d15b1   docker-fpm-frr:latest             "/usr/bin/docker_ini…"   2 hours ago   Up 2 minutes             bgp1
4e35cf0f962f   docker-syncd-brcm-dnx:latest      "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             syncd0
14f9109c93cc   docker-teamd:latest               "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             teamd0
f77fd9c53cd7   docker-syncd-brcm-dnx:latest      "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             syncd1
57209995a85b   docker-teamd:latest               "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             teamd1
fd030291d9fa   docker-orchagent:latest           "/usr/bin/docker-ini…"   2 hours ago   Up 2 minutes             swss1
cc0377006913   docker-eventd:latest              "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             eventd
ae7c871d4e37   docker-database:latest            "/usr/local/bin/dock…"   2 hours ago   Up 2 minutes             database1
aeb234180f94   docker-database:latest            "/usr/local/bin/dock…"   2 hours ago   Up 2 minutes             database0
f0ed626a8152   docker-database:latest            "/usr/local/bin/dock…"   2 hours ago   Up 2 minutes             database
admin@ixre-egl-board40:~$ cat empty.json
[]

admin@ixre-egl-board40:~$ sudo config apply-patch empty.json -d
** DRY RUN EXECUTION **
Patch Applier: Patch application starting.
Patch Applier: Patch: []
Patch Applier: Getting current config db.
Patch Applier: Simulating the target full config after applying the patch.
Patch Applier: Validating all JsonPatch operations are permitted on the specified fields
Patch Applier: Validating target config does not have empty tables, since they do not show up in ConfigDb.
Patch Applier: Sorting patch updates.
Patch Applier: The patch was converted into 0 changes.
Patch Applier: Applying 0 changes in order.
Patch Applier: Verifying patch updates are reflected on ConfigDB.
Patch Applier: Patch application completed.
Patch applied successfully.
admin@ixre-egl-board40:~$ docker ps
CONTAINER ID   IMAGE                          COMMAND                  CREATED       STATUS         PORTS     NAMES
3a8ddbc204ad   docker-fpm-frr:latest          "/usr/bin/docker_ini…"   2 hours ago   Up 2 minutes             bgp0
e553479d15b1   docker-fpm-frr:latest          "/usr/bin/docker_ini…"   2 hours ago   Up 2 minutes             bgp1
4e35cf0f962f   docker-syncd-brcm-dnx:latest   "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             syncd0
f77fd9c53cd7   docker-syncd-brcm-dnx:latest   "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             syncd1
cc0377006913   docker-eventd:latest           "/usr/local/bin/supe…"   2 hours ago   Up 2 minutes             eventd
ae7c871d4e37   docker-database:latest         "/usr/local/bin/dock…"   2 hours ago   Up 2 minutes             database1
aeb234180f94   docker-database:latest         "/usr/local/bin/dock…"   2 hours ago   Up 2 minutes             database0
f0ed626a8152   docker-database:latest         "/usr/local/bin/dock…"   2 hours ago   Up 2 minutes             database
admin@ixre-egl-board40:~$

```

**Comment by arlakshm** [comment]
@JunhongMao and @mlok-nokia, as discussed offline with update the PR will latest proposal.

**Comment by JunhongMao** [comment]
These two PRs 
https://github.com/sonic-net/sonic-buildimage/pull/18578
https://github.com/nokia/sonic-platform/pull/6 

have been replaced by the below new two PRS:

https://github.com/sonic-net/sonic-utilities/pull/3283
https://github.com/sonic-net/sonic-platform-daemons/pull/475

File: files/build_templates/chassis-module.service (hunk 1 above)
Comment (mlok-nokia) [reply files/build_templates/chassis-module.service:4]: No longe applicable

**Pull request closed without merging**